stl/inc/xutility

```diff
@@ -2176,12 +2181,19 @@ namespace ranges {
 
             template <class _Ty>
             _NODISCARD static _CONSTEVAL _Choice_t<_St> _Choose() noexcept {
-                if constexpr (is_array_v<remove_reference_t<_Ty>>) {
-                    return {_St::_Array, true};
-                } else if constexpr (_Has_member<_Ty&>) {
-                    return {_St::_Member, noexcept(_Fake_decay_copy(_STD declval<_Ty&>().end()))};
-                } else if constexpr (_Has_ADL<_Ty&>) {
-                    return {_St::_Non_member, noexcept(_Fake_decay_copy(end(_STD declval<_Ty&>())))};
+                _STL_INTERNAL_CHECK(is_lvalue_reference_v<_Ty>);
+                using _UnRef = remove_reference_t<_Ty>;
+
+                if constexpr (is_array_v<_UnRef>) {
+                    if constexpr (extent_v<_UnRef> != 0) {
+                        return {_St::_Array, true};
+                    } else {
+                        return {_St::_None};
+                    }

```

Comment:
I remember implementing `std::is_unbounded_array` and `std::is_bounded_array`, which would be a better fit i believe:

```suggestion
    if constexpr (is_bounded_array_v<remove_reference_t<_Ty>>) {
        return {_St::_Array, true};
    } else if constexpr (is_unbounded_array_v<remove_reference_t<_Ty>>) {
        return {_St::_None};
    } else if constexpr (_Has_member<_Ty>) 
```

Reply:
That's how I implemented this initially. It instantiates:
* bounded array case: one variable template
* unbounded array case: two variable templates
* non-array cases: two variable templates
vs. what's in the PR:
* bounded array case: two variable templates minus one since `extent_v` is reused in `operator()` = net one variable template
* unbounded array case: two variable templates
* non-array cases: one variable templates
I chose the latter since it's it least as efficient in every case, and more efficient in the more common bounded array and non-array cases.

Reply:
And today I talked with a coworker about the incredible length you go to get every little bit out of the compiler.
:+1: